File: ProjectEuler/101-150/S110.cs
using System;
using System.Linq;
using System.Numerics;

namespace ProjectEuler
{
    public class S110 : ISolution
    {
        private readonly int MAX = 4_000_000;
        private readonly int[] Primes = new int[] { 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37 };
        private readonly int[] MaxPower = new int[] { 9, 5, 4, 2, 2, 2, 2, 2, 2, 2, 2, 2 };
        // 18,700,260,099,721,200
        private BigInteger min = long.MaxValue;

        public string GetAnswer()
        {
            Check(new int[] { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 }, 0);
            return min.ToString();
        }

        private void Check(int[] power, int index)
        {
            if (index == power.Length)
            {
                int count = 1;
                foreach (var item in power)
                {
                    count *= (1 + item * 2);
                }
                count++;
                count /= 2;

                if (count > MAX)
                {
                    var n = Primes.Zip(power, (i, j) => (long)Math.Pow(i, j))
                        .Select(l => (BigInteger)l).Aggregate(BigInteger.Multiply);
                    if (n < min)
                    {
                        min = n;
                    }
                }
            }
            else
            {
                for (int i = 0; i <= MaxPower[index]; i++)
                {
                    var copy = power.ToArray();
                    copy[index] = i;
                    Check(copy, index + 1);
                }
            }
        }
    }
}

File: src/101-150/S110.cc
#include "S110.h"
#include "../BigInt.h"
#include "../Utils.h"
#include <climits>
#include <cmath>
#include <iostream>
#include <vector>

using namespace std;

namespace ProjectEuler
{
std::string S110::GetAnswer()
{
	// Initialize min with a large upper bound:
	// Product of primes up to solution.
	// Let's just run once with MaxPower to set a valid min.
	min = 0; // Invalid
	bool first = true;

	// Check with initial powers = 0
	std::vector<int> powers(Primes.size(), 0);
	Check(powers, 0);

	return min.toString();
}

void S110::Check(std::vector<int> power, int index)
{
	if (index == Primes.size())
	{
		long long count = 1;
		for (int item : power)
		{
			count *= (1 + item * 2);
		}

		long long solutions = (count + 1) / 2;

		if (solutions > MAX)
		{
			BigInt n = 1;
			for (size_t i = 0; i < power.size(); ++i)
			{
				if (power[i] > 0)
				{
					n = n * Utils::Pow(Primes[i], power[i]);
				}
			}

			if (min == 0 || n < min)
			{
				min = n;
			}
		}
	}
	else
	{
		for (int i = 0; i <= MaxPower[index]; i++)
		{
			// Pruning: powers should be non-increasing for optimal n
			if (index > 0 && i > power[index - 1])
			{
				break;
			}

			std::vector<int> copy = power;
			copy[index] = i;
			Check(copy, index + 1);
		}
	}
}

} // namespace ProjectEuler
